### PR TITLE
bpo-36540: PEP 570 -- Implementation

### PR DESCRIPTION
**Pull request by pablogsal**
implementation of PEP 570 -- Python Positional-Only Parameters:

:warning: **This Pull Request contains only the implementation of PEP570, the documentation will be done in a different PR**.

https://www.python.org/dev/peps/pep-0570/

https://bugs.python.org/issue36540

CC: @mariocj89

### CODE REVIEW

File: Lib/test/test_functools.py
Comment (pganssle) [inline Lib/test/test_functools.py]: I don't think this should work with a keyword - it wouldn't work for current positional-only arguments implemented in the C, I don't think there's an easy way to *make* it work for those.

I do think it would be very useful if `partial` could curry positional arguments other than the first one, though. There are a lot of situations where that could come in handy. Maybe this is good impetus for trying to solve that problem?

File: Python/Python-ast.c
Comment (pganssle) [inline Python/Python-ast.c:3962]: Per PEP 7, you should have curly braces after all control statements, even where optional.

It may be that I'm mis-reading this diff and you're not actually *adding* non-compliant code, but I think it's worth a refactor-as-you-go fix for this.

File: Python/Python-ast.c
Comment (pganssle) [inline Python/Python-ast.c:8337]: Another missing braces here, and on line 8333.

File: Python/ast.c
Comment (pganssle) [inline Python/ast.c]: Braces again, I'd also probably add them to the line above even though you didn't add it.

File: Python/marshal.c
Comment (pganssle) [inline Python/marshal.c:1350]: Another  place that needs curly braces.

I'm going to stop commenting on these to cut down on spam, if you want me to point them out I can do another pass where I note all of them.

**Review by pganssle — CHANGES_REQUESTED** [review]
Some comments here from a very cursory pass over this.

Is the documentation for this intended to be part of a separate PR? Either way, feel free to ping me for a review when it's available.

File: Python/Python-ast.c
Comment (pablogsal) [reply Python/Python-ast.c:3962]: This file is automatically generated

File: Python/Python-ast.c
Comment (pablogsal) [reply Python/Python-ast.c:8337]: Check out the previous comment

File: Python/marshal.c
Comment (pablogsal) [reply Python/marshal.c:1350]: In this case I have matched the style of the files to not make the diff huge. The PR is already gigantic so I prefer to fix it later (if fixed at all).

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: I would return `posonlyargs` before `args`.

Comment (serhiy-storchaka) [reply Lib/inspect.py]: Actually, `_getfullargs()` is only used in `getargs()`, so it may be better to inline it.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py:1185]: This is breaking change. `getfullargspec()` was added to avoid breaking old `getargspec()`. You need to add a new function (`getfullerargspec()`?) and deprecate `getfullargspec()`.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py:1217]: Since all these parameters are positional-or-keyword, inserting a new parameter in the middle is breaking change. You can add it only to the end. Or add a new function.

File: Python/marshal.c
Comment (pganssle) [reply Python/marshal.c:1350]: You don't need to change all the braces in the file, but IIRC pure-refactoring PRs are not generally accepted for CPython, so you're supposed to "clean up as you go".

I don't see any particular reason to try to maintain stylistic consistency around braces use even within a function, so I don't think it would make the diff much bigger if you just follow PEP 7 within the code you are adding or changing.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: You can use `[*posonlyargs, *args]`.

File: Lib/test/test_parser.py
Comment (serhiy-storchaka) [inline Lib/test/test_parser.py:244]: Test also optional position-only parameters.

File: Lib/test/test_positional_only_arg.py
Comment (serhiy-storchaka) [inline Lib/test/test_positional_only_arg.py]:
```
return (a, b)
```
or
```
return a, b
```

File: Lib/test/test_positional_only_arg.py
Comment (serhiy-storchaka) [inline Lib/test/test_positional_only_arg.py]: What is optional here?

File: Lib/test/test_positional_only_arg.py
Comment (serhiy-storchaka) [inline Lib/test/test_positional_only_arg.py:21]: You can use `test.support.check_syntax_error()`. Just add
```
from test.support import check_syntax_error
```
in the class body.

File: Lib/test/test_type_comments.py
Comment (serhiy-storchaka) [inline Lib/test/test_type_comments.py]: Why this is needed?

File: Python/marshal.c
Comment (pablogsal) [reply Python/marshal.c:1350]: There is a big chain of conditionals in this function (and the file is fairly inconsistent). I still prefer not to do that refactoring in this PR. :(

File: Objects/codeobject.c
Comment (serhiy-storchaka) [inline Objects/codeobject.c]: Missed space.

File: Objects/codeobject.c
Comment (serhiy-storchaka) [inline Objects/codeobject.c]: Missed space after comma.

File: Objects/codeobject.c
Comment (serhiy-storchaka) [inline Objects/codeobject.c]:
```suggestion
    if (!eq) goto unequal;
```

File: Python/ast.c
Comment (serhiy-storchaka) [inline Python/ast.c:1492]: The comment above should be updated.

File: Python/marshal.c
Comment (pganssle) [reply Python/marshal.c:1350]: I was just suggesting that in this part you added, you do:

```C
if (PyErr_Occurred()) {
    goto code_error;
}
posonlyargcount = (int)r_long(p);
```

You don't need to refactor all the other ones in the chain. If anything I think consistency hurts here because it discourages people who touch this file from adding the braces.

Comment (pablogsal) [reply Python/marshal.c:1350]: Ok, done in a85a43f

File: Lib/test/inspect_fodder2.py
Comment (vstinner) [inline Lib/test/inspect_fodder2.py]: Can you add spaces? PEP 8.

File: Lib/modulefinder.py
Comment (vstinner) [inline Lib/modulefinder.py:622]: This construction is not future-proof, it breaks each time we add an argument to the Code constructor. Maybe we need a .replace() method to replace a few fields and copies other, similar to datetime.datetime.replace() (datetime.datime is also immutable).

That's a note unrelated to your PR ;-)

File: Lib/test/test_inspect.py
Comment (vstinner) [inline Lib/test/test_inspect.py]: Why is it commented?

File: Lib/test/test_positional_only_arg.py
Comment (vstinner) [inline Lib/test/test_positional_only_arg.py]: Please sort imports.

File: Python/ceval.c
Comment (vstinner) [inline Python/ceval.c]: Since it is not a direct copy of the code object, can you remove co_ prefix from the variable name?

File: Python/importlib_zipimport.h
Comment (vstinner) [inline Python/importlib_zipimport.h:1]: Hum, we should mark these generated files as binary in .gitattributes to try to hide them in reviews. I dont ask you to do that in your PR.

**Review by vstinner — CHANGES_REQUESTED** [review]
You modified marshal, so you must increase pyc magic number in Lib/importlib. Apart that that, I dont see any blocker issue.

It seems like you modified inspect functions which impacts the API. That should be properly documented. I'm fine with documenting in a separated PR.

Same remark for Code constructor: API changed.

**Comment by pablogsal** [comment]
Thanks, @vstinner for the review!

>You modified marshal, so you must increase pyc magic number in Lib/importlib. Apart that that, I dont see any blocker issue.

I have increased the magic number to `3410` (from 3401) and I have addressed the rest of your feedback in the latest commit.

> It seems like you modified inspect functions which impacts the API. That should be properly documented. I'm fine with documenting in a separated PR.
>Same remark for Code constructor: API changed.

Yup, I plan to address all documentation in a separate PR to keep this a bit more manageable for me and for the reviewers.

I have made the requested changes; please review again

**Review by vstinner — APPROVED** [review]
LGTM. I didn't look at the BNF Grammar, I'm unable to review that. As soon as your change doesn't break other tests, I'm fine with Grammar changes :-)

@emilyemorehouse, @gvanrossum: Would you mind to review it? If you don't have the bandwidth, I would suggest to merge this change soon, to get more time before beta1 release, to play with the implementation and detect any corner case.

**Comment by serhiy-storchaka** [comment]
Please let me to finish a review.

**Comment by pablogsal** [comment]
@serhiy-storchaka Gentle ping for the review.

As has been commented here and in other PRs, I would like to merge this before Alpha 4 if possible, so this can have more exposure and tested outside. We can do this without the inspect changes that you mentioned in the issue.

Please, feel free to push directly to my branch if you want to change something or you spot some errors so we can reduce the time required for the review process.

**Comment by vstinner** [comment]
About changes of the inspect module and backward-compatibility, let me have a look.

* getargs() and getargspec() API is unchanged
* getfullargspec() gets a new posonlyargs field in its FullArgSpec type: getfullargspec()[5] now returns posonlyargs rather than kwonlyargs
* formatargspec() gets a new 'posonlyargs' positional-or-keyword argument between "defaults" and "kwonlyargs": this change looks backward incompatible.

Maybe something should be done for getfullargspec() and formatargspec().

**Comment by vstinner** [comment]
getfullargspec() was deprecated but then became undeprecated again... https://bugs.python.org/issue20438

**Comment by vstinner** [comment]
My advice is to merge this PR as it is. Maybe it's not perfect: that's ok. The important part is to get the new syntax as soon as possible, to get enough time to play with it before Python 3.8.0 final. I expect that subtle issue will be found, and I prefer to find them before the final rather than after.

I suggest to leave inspect unchanged *but* open a new issue on bugs.python.org or open a thread on python-dev to decide how to handle the backward-compatibility. IMHO there is no good solution, only compromises. You cannot add support for positional-only arguments and leave the API unchanged. I'm talking about getargs(), getargspec(), getfullargspec() and formatargspec() functions. signature() is fine if I understood correctly.

**Comment by pablogsal** [comment]
I have removed the changes in `inspect.py` that affect deprecated functions (`getargspec`...) and we raise if the user passes something that has positional only arguments.

**Comment by gvanrossum** [comment]
I'm not crazy about raising. Maybe just pretend the '/' isn't there?

**Comment by pablogsal** [comment]
> I'm not crazy about raising. Maybe just pretend the '/' isn't there?

We already raise if you pass keyword only for that API

**Comment by vstinner** [comment]
> We already raise if you pass keyword only for that API

That was my proposal. IMHO it's the safest option for "legacy" functions of the inspect module, to ensure that we don't introduce any subtle backward-incompatible change.

I modified the warnings module to add a new "source" parameter at the end. It took me 3 years to fix regressions on corner cases... The latest one was https://github.com/python/cpython/commit/be7c460fb50efe3b88a00281025d76acc62ad2fd See https://bugs.python.org/issue26568 and https://bugs.python.org/issue35178

I tried to have the backward-compatibility in mind when I modified the warnings module, but there are always subtle issues...

IMHO getargs(), getargspec(), getfullargspec() should be deprecated in favor of signature() which is future-proof. But Nick Coghlan undeprecated these functions in 2017 https://bugs.python.org/issue20438 whereas these were deprecated since 2015. I didn't understand the subtle issues.

Again, that's why I suggest to open an issue afterwards to discuss the subtle inspect API :-)

**Comment by gvanrossum** [comment]
Oh, then disregard what I said.

**Comment by pablogsal** [comment]
As a side comment, Cython is already adding support for this PEP :tada:  :

https://github.com/cython/cython/pull/2927

**Comment by vstinner** [comment]
@pablogsal: Please merge your PR.

Serhiy Storchaka complained about backward incompatible changes on the inspect module:
https://bugs.python.org/issue36540#msg340246

But I made the same complain and you fixed it. IHMO the change is now good enough to be merged.

Serhiy said nothing for 2 weeks. alpha4 release is today. I would prefer to get this implementation merged as soon as possible: as Lukasz if I understand correctly.

**Comment by serhiy-storchaka** [comment]
My main complain was about the inspect module. I have no time to make a review of other parts, but I do not expect serious problems here. Feel free to commit. We can fix minor issues after merging if there are any.

**Comment by vstinner** [comment]
@serhiy-storchaka:
> My main complain was about the inspect module. I have no time to make a review of other parts, but I do not expect serious problems here. Feel free to commit. We can fix minor issues after merging if there are any.

Did you see that the PR has been modified to no longer touch the existing API, except of signature(), but raise an exception of the "legacy functions" are called on a function which has positional-only arguments? https://github.com/python/cpython/pull/12701#issuecomment-486433635

File: Lib/inspect.py
Comment (pablogsal) [reply Lib/inspect.py:1217]: I suspect that there may be a longer discussion for these cases so I opened https://bugs.python.org/issue36751 to track it and decide there.

File: Lib/inspect.py
Comment (pablogsal) [reply Lib/inspect.py:1185]: I suspect that there may be a longer discussion for these cases so I opened https://bugs.python.org/issue36751 to track it and decide there.

**Comment by pablogsal** [comment]
> Hi! The buildbot x86 Windows7 3.x has failed when building commit 8c77b8c.

False positive. It is some problem in the multiprocessing module.

File: Include/code.h
Comment (Ark-kun) [inline Include/code.h:106]: Adding a required argument to the code type constructor will breaking every library that creates `types.CodeType` instances. Are you sure this is a good idea?

Comment (Ark-kun) [reply Include/code.h:106]: Looks like it has already started: https://www.google.com/search?q=TypeError%3A+code%28%29+%22takes+at+least+14+arguments+%2813+given%29%22

Comment (vstinner) [reply Include/code.h:106]: You missed the end of the story:

```
commit 4a2edc34a405150d0b23ecfdcb401e7cf59f4650
Author: Pablo Galindo <Pablogsal@gmail.com>
Date:   Mon Jul 1 11:35:05 2019 +0100

    bpo-37221: Add PyCode_NewWithPosOnlyArgs to be used internally and set PyCode_New as a compatibility wrapper (GH-13959)
    
    Add PyCode_NewEx to be used internally and set PyCode_New as a compatibility wrapper
```

Comment (pablogsal) [reply Include/code.h:106]: Also, `types.CodeType`'s constructor is considered private. We now even have this warning in the docs (https://docs.python.org/3.8/library/types.html#standard-interpreter-types):

> If you instantiate any of these types, note that signatures may vary between Python versions.

Additionally, you have now available `types.CodeType.replace` in case you can to create a copy with some fields changed.

Comment (Ark-kun) [reply Include/code.h:106]: Thank you for the explanation!
I hope in future we'll have a more stable API for dynamic function generation.

Comment (ManifoldFR) [reply Include/code.h:106]: This breaks `pickle.loads` for functions constructed as
```python
def constant_fn(val: float) -> Callable:
    def func(_):
        return val
    
    return func
fun = constant_fn(1.)
```
and pickled using `cloudpickle` under Python 3.7; these can't be loaded back up under Python 3.8.
Should this be addressed downstream in cloudpickle ?

Comment (pablogsal) [reply Include/code.h:106]: > Should this be addressed downstream in cloudpickle ?

Yes, they need to update whatever code they are using to create code objects as code objects in Python3.8 require an extra field. Please, check these sections in the docs:

https://docs.python.org/3/library/types.html#standard-interpreter-types

> If you instantiate any of these types, note that signatures may vary between Python versions.

Also, check this new C-API function:

https://docs.python.org/3/c-api/code.html#c.PyCode_NewWithPosOnlyArgs

and this new convenience function:

https://docs.python.org/3/library/types.html#types.CodeType.replace

Comment (ManifoldFR) [reply Include/code.h:106]: Alright, I'll file an issue there. Thanks.

**Pull request merged**
